File: Client-Service/client/src/services/api/userCustomerMappingService.ts
import api from "../api";
import { ApiResponse } from "../api";

// User-Customer Mapping DTOs
export interface UserCustomerMapping {
  mappingId: number;
  userId: number;
  customerId: number;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateMappingRequest {
  userId: number;
  customerId: number;
}

export interface MappingResponse {
  success: boolean;
  message: string;
  mapping?: UserCustomerMapping;
}

// User-Customer Mapping API Service
export const userCustomerMappingService = {
  // Create mapping between user and customer
  createMapping: async (mappingData: CreateMappingRequest): Promise<UserCustomerMapping> => {
    const response = await api.post<UserCustomerMapping>('user-customer-mapping/create', mappingData);
    return response.data;
  },

  // Get mapping by user ID
  getMappingByUserId: async (userId: number): Promise<UserCustomerMapping | null> => {
    const response = await api.get<UserCustomerMapping>(`user-customer-mapping/user/${userId}`);
    return response.data;
  },

  // Get mapping by customer ID
  getMappingByCustomerId: async (customerId: number): Promise<UserCustomerMapping | null> => {
    const response = await api.get<UserCustomerMapping>(`user-customer-mapping/customer/${customerId}`);
    return response.data;
  },

  // Get all mappings for a user
  getAllMappingsForUser: async (userId: number): Promise<UserCustomerMapping[]> => {
    const response = await api.get<UserCustomerMapping[]>(`user-customer-mapping/user/${userId}/all`);
    return response.data;
  },

  // Get all mappings for a customer
  getAllMappingsForCustomer: async (customerId: number): Promise<UserCustomerMapping[]> => {
    const response = await api.get<UserCustomerMapping[]>(`user-customer-mapping/customer/${customerId}/all`);
    return response.data;
  },

  // Update mapping status
  updateMappingStatus: async (mappingId: number, status: string): Promise<UserCustomerMapping> => {
    const response = await api.put<UserCustomerMapping>(`user-customer-mapping/${mappingId}/status`, {
      status
    });
    return response.data;
  },

  // Delete mapping
  deleteMapping: async (mappingId: number): Promise<void> => {
    await api.delete(`user-customer-mapping/${mappingId}`);
  },

  // Delete mapping by user and customer
  deleteMappingByUserAndCustomer: async (userId: number, customerId: number): Promise<void> => {
    await api.delete(`user-customer-mapping/user/${userId}/customer/${customerId}`);
  },

  // Get customer ID for current user
  getCustomerIdForCurrentUser: async (): Promise<number | null> => {
    const response = await api.get<number>('user-customer-mapping/current/customer-id');
    return response.data;
  },

  // Get user ID for customer
  getUserIdForCustomer: async (customerId: number): Promise<number | null> => {
    const response = await api.get<number>(`user-customer-mapping/customer/${customerId}/user-id`);
    return response.data;
  },

  // Verify if user has access to customer
  verifyUserCustomerAccess: async (userId: number, customerId: number): Promise<boolean> => {
    const response = await api.get<boolean>(`user-customer-mapping/verify/${userId}/${customerId}`);
    return response.data;
  }
};

File: Customer-Service/customer/src/main/java/com/bank/customer/service/UserCustomerMappingService.java
package com.bank.customer.service;

import com.bank.customer.model.UserCustomerMapping;
import com.bank.customer.repository.UserCustomerMappingRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * User-Customer Mapping Service - STRICT SECURITY MODEL
 * Manages the relationship between authentication users and business customers
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class UserCustomerMappingService {
    
    private final UserCustomerMappingRepository mappingRepository;
    
    /**
     * Create mapping between user and customer
     */
    public UserCustomerMapping createMapping(Long userId, Long customerId) {
        log.info("Creating user-customer mapping - UserId: {}, CustomerId: {}", userId, customerId);
        
        UserCustomerMapping mapping = UserCustomerMapping.builder()
                .userId(userId)
                .customerId(customerId)
                .status("ACTIVE")
                .createdAt(java.time.Instant.now())
                .updatedAt(java.time.Instant.now())
                .build();
        
        return mappingRepository.save(mapping);
    }
    
    /**
     * Get mapping by user ID
     */
    public Optional<UserCustomerMapping> findByUserId(Long userId) {
        return mappingRepository.findByUserId(userId);
    }
    
    /**
     * Get mapping by customer ID
     */
    public Optional<UserCustomerMapping> findByCustomerId(Long customerId) {
        return mappingRepository.findByCustomerId(customerId);
    }
    
    /**
     * Check if mapping exists by user ID and customer ID
     */
    public boolean existsByUserIdAndCustomerId(Long userId, Long customerId) {
        return mappingRepository.existsByUserIdAndCustomerId(userId, customerId);
    }
    
    /**
     * Get all mappings for a user
     */
    public List<UserCustomerMapping> findAllByUserId(Long userId) {
        return mappingRepository.findAllByUserId(userId);
    }
    
    /**
     * Get all mappings for a customer
     */
    public List<UserCustomerMapping> findAllByCustomerId(Long customerId) {
        return mappingRepository.findAllByCustomerId(customerId);
    }
    
    /**
     * Update mapping status
     */
    public UserCustomerMapping updateStatus(Long mappingId, String status) {
        UserCustomerMapping mapping = mappingRepository.findById(mappingId)
                .orElseThrow(() -> new RuntimeException("Mapping not found: " + mappingId));
        
        mapping.setStatus(status);
        mapping.setUpdatedAt(java.time.Instant.now());
        
        return mappingRepository.save(mapping);
    }
    
    /**
     * Delete mapping
     */
    public void deleteMapping(Long mappingId) {
        UserCustomerMapping mapping = mappingRepository.findById(mappingId)
                .orElseThrow(() -> new RuntimeException("Mapping not found: " + mappingId));
        
        mappingRepository.delete(mapping);
        log.info("Deleted user-customer mapping: {}", mappingId);
    }
    
    /**
     * Delete mapping by user and customer
     */
    public void deleteMappingByUserIdAndCustomerId(Long userId, Long customerId) {
        Optional<UserCustomerMapping> mapping = mappingRepository.findByUserIdAndCustomerId(userId, customerId);
        
        mapping.ifPresent(m -> {
            mappingRepository.delete(m);
            log.info("Deleted user-customer mapping - UserId: {}, CustomerId: {}", userId, customerId);
        });
    }
}
